From d665a7510e42a124d240eb453c224932a5c876e7 Mon Sep 17 00:00:00 2001
Message-Id: <d665a7510e42a124d240eb453c224932a5c876e7.1424442690.git.erack@redhat.com>
From: Eike Rathke <erack@redhat.com>
Date: Fri, 20 Feb 2015 15:11:36 +0100
Subject: [PATCH] Resolves: tdf#89500 catch ISO 8601 datetime in all locales
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="------------erAck-patch-parts"

This is a multi-part message in MIME format.
--------------erAck-patch-parts
Content-Type: text/plain; charset=UTF-8; format=fixed
Content-Transfer-Encoding: 8bit


Change-Id: I906ebdc6aa5c365df98a29e9e0f319f89f6833c6
(cherry picked from commit 5d572e673dec3396487b14bb1f9e86aa8ab86786)
---
 svl/source/numbers/zforfind.cxx | 33 ++++++++++++++++++++++++++-------
 1 file changed, 26 insertions(+), 7 deletions(-)


--------------erAck-patch-parts
Content-Type: text/x-patch; name="0001-Resolves-tdf-89500-catch-ISO-8601-datetime-in-all-lo.patch"
Content-Transfer-Encoding: 8bit
Content-Disposition: attachment; filename="0001-Resolves-tdf-89500-catch-ISO-8601-datetime-in-all-lo.patch"

diff --git a/svl/source/numbers/zforfind.cxx b/svl/source/numbers/zforfind.cxx
index 5aee40d..f709cfe 100644
--- a/svl/source/numbers/zforfind.cxx
+++ b/svl/source/numbers/zforfind.cxx
@@ -2439,16 +2439,35 @@ bool ImpSvNumberInputScan::ScanMidString( const OUString& rString,
                     SkipBlanks( rString, nPos );
                 }
             }
-            else if (nStringPos == 5 && nPos == 0 && rString.getLength() == 1 &&
-                     rString[ 0 ] == 'T' && MayBeIso8601())
+            else if (nPos == 0 && rString.getLength() == 1 && MayBeIso8601())
             {
-                // ISO 8601 combined date and time, yyyy-mm-ddThh:mm
-                ++nPos;
+                if (nStringPos == 5 && rString[0] == 'T')
+                {
+                    // ISO 8601 combined date and time, yyyy-mm-ddThh:mm
+                    ++nPos;
+                }
+                else if (nStringPos == 7 && rString[0] == ':')
+                {
+                    // ISO 8601 combined date and time, the time part; we reach
+                    // here if the locale's separator is not ':' so it couldn't
+                    // be detected above in the time block.
+                    if (nAnzNums >= 5)
+                        eScannedType = NUMBERFORMAT_DATETIME;
+                    ++nPos;
+                }
             }
             break;
-#if NF_RECOGNIZE_ISO8601_TIMEZONES
         case NUMBERFORMAT_DATETIME:
-            if (nPos == 0 && rString.getLength() == 1 && nStringPos >= 9 && MayBeIso8601())
+            if (nPos == 0 && rString.getLength() == 1 && MayBeIso8601())
+            {
+                if (nStringPos == 9 && rString[0] == ':')
+                {
+                    // ISO 8601 combined date and time, the time part continued.
+                    ++nPos;
+                }
+            }
+#if NF_RECOGNIZE_ISO8601_TIMEZONES
+            else if (nPos == 0 && rString.getLength() == 1 && nStringPos >= 9 && MayBeIso8601())
             {
                 // ISO 8601 timezone offset
                 switch (rString[ 0 ])
@@ -2475,8 +2494,8 @@ bool ImpSvNumberInputScan::ScanMidString( const OUString& rString,
                     break;
                 }
             }
-            break;
 #endif
+            break;
         }
     }
 

--------------erAck-patch-parts--


